framework/src/components/system/system.js

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const os = require('os');
+const crypto = require('crypto');
+const semver = require('semver');
+
+/**
+ * Main system methods. Initializes library with scope content and headers:
+ * - os
+ * - version
+ * - wsPort
+ * - httpPort
+ * - minVersion
+ * - protocolVersion
+ * - height
+ * - nethash
+ * - broadhash
+ * - nonce
+ *
+ * @class
+ * @requires crypto
+ * @requires os
+ * @requires semver
+ * @param {Object} options - System options
+ * @param {Object} logger
+ */
+class System {
+	constructor(config, logger, storage) {
+		this.logger = logger;
+		this.storage = storage;
+		this.headers = {
+			os: os.platform() + os.release(),
+			version: config.version,
+			wsPort: config.wsPort,
+			httpPort: config.httpPort,
+			minVersion: config.minVersion,
+			protocolVersion: config.protocolVersion,
+			height: 1,
+			nethash: config.nethash,
+			broadhash: config.nethash,
+			nonce: config.nonce,
+		};
+	}
+
+	/**
+	 * Checks nethash (network) compatibility.
+	 *
+	 * @param {string} nethash
+	 * @returns {boolean}
+	 */
+	networkCompatible(nethash) {
+		if (!nethash) {
+			return false;
+		}
+		return this.headers.nethash === nethash;
+	}
+
+	/**
+	 * Checks version compatibility from input param against private values.
+	 *
+	 * @param {string} version
+	 * @returns {boolean}
+	 */
+	versionCompatible(version) {
+		if (!version) {
+			return false;
+		}
+		return semver.gte(version, this.headers.minVersion);
+	}
+
+	/**
+	 * Checks protocol version compatibility from input param against
+	 * private values.
+	 *
+	 * @param protocolVersion
+	 * @returns {boolean}
+	 */
+	protocolVersionCompatible(protocolVersion) {
+		if (!protocolVersion) {
+			return false;
+		}
+		const peerHard = parseInt(protocolVersion[0]);
+		const myHard = parseInt(this.headers.protocolVersion[0]);
+		return myHard === peerHard && peerHard >= 1;
+	}
+
+	/**
+	 * Checks nonce (unique app id) compatibility- compatible when different than given.
+	 *
+	 * @param nonce
+	 * @returns {boolean}
+	 */
+	nonceCompatible(nonce) {
+		if (!nonce) {
+			return false;
+		}
+		return nonce && this.headers.nonce !== nonce;
+	}
+
+	/**
+	 * Updates private broadhash and height values.
+	 *
+	 * @param {function} cb - Callback function
+	 * @returns {setImmediateCallback} cb, err
+	 */
+	update(cb) {
+		return this.storage.entities.Block.get(
```

Comment:
This function is returning promise as well as accepting a callback. Either we should refactor to not use callback or we should not return the Promise.

Reply:
Already discussed #2904